drivers/media/i2c/mira220.c

```diff
@@ -0,0 +1,2481 @@
+
+#define MIRA220_OTP_CMD_REG CCI_REG8(0x0080)
+#define MIRA220_OTP_CMD_UP 0x4
+#define MIRA220_OTP_CMD_DOWN 0x8
+
+#define MIRA220_GLOB_NUM_CLK_CYCLES 1928
+
+#define MIRA220_SUPPORTED_XCLK_FREQ 38400000
+
+// Default exposure is adjusted to mode with smallest height
+#define MIRA220_DEFAULT_EXPOSURE \
+	100 //(MIRA220_MIN_V_SIZE + MIRA220_MIN_VBLANK - MIRA220_GLOB_NUM_CLK_CYCLES / MIRA220_MIN_ROW_LENGTH) //TODO
+#define MIRA220_EXPOSURE_MIN 1
+
+// Power on function timing
+#define MIRA220_XCLR_MIN_DELAY_US 100000
+#define MIRA220_XCLR_DELAY_RANGE_US 30
+
+// Outdated. See below.
+// pixel_rate = link_freq * 2 * nr_of_lanes / bits_per_sample
+// 1.0Gb/s * 2 * 2 / 12 = 357913941
+// #define MIRA220_PIXEL_RATE		(357913941)
+
+// Mira220 PIXEL_RATE is derived from ROW_LENGTH. See datasheet Section 9.2.
+// ROW_LENGTH is set by registers: 0x102B, 0x102C. Unit is number of CLK_IN cycles.
+// PIXEL_RATE = 1000000000 * WIDTH / (ROW_LENGTH * CLK_IN_PERIOD_NS)
+// ROW_LENGTH_1600x1400_1000GBS=300
+// ROW_LENGTH_640x480_1000GBS=450
+// CLK_IN_PERIOD_NS = 1.0 s / 38.4 Mhz = 26.04 ns
+// MIRA220_PIXEL_RATE = 1000000000 * 1600 / (300 * 26.04) = 204813108
+// MIRA220_PIXEL_RATE = 1000000000 * 640 / (450 * 26.04) = 54616828
+#define MIRA220_PIXEL_RATE 384000000 //384M (x10)
+// Row time in microseconds. Not used in driver, but used in libcamera cam_helper.
+// ROW_TIME_US = ROW_LENGTH * CLK_IN_PERIOD_NS / 1000
+// MIRA220_ROW_TIME_1600x1400_1000GBS_US=(300*26.04/1000)=7.8us
+// MIRA220_ROW_TIME_640x480_1000GBS_US=(450*26.04/1000)=11.7us
+
+/* Should match device tree link freq */
+#define MIRA220_DEFAULT_LINK_FREQ 750000000
+
+/* Trick the libcamera with achievable fps via hblank */
+
+/* Formular in libcamera to derive TARGET_FPS:
+  * TARGET_FPS=1/((1/MIRA220_PIXEL_RATE)*(WIDTH+HBLANK)*(HEIGHT+MIRA220_MIN_VBLANK))
+  * Example: 640x480 with HBLANK=0 and MIRA220_MIN_VBLANK=13
+  * TARGET_FPS=1/((1/54616828)*640*(480+13))=173
+  * Example: 1600x1400 with HBLANK=0 and MIRA220_MIN_VBLANK=13
+  * TARGET_FPS=1/((1/204813108)*1600*(1400+13))=91
+  * 
+  * Inverse the above formula to derive HBLANK from TARGET_FPS:
+  * HBLANK=1/((1/MIRA220_PIXEL_RATE)*TARGET_FPS*(HEIGHT+MIRA220_MIN_VBLANK))-WIDTH
+  * Example with TARGET_FPS of 120 fps for 640x480
+  * HBLANK=1/((1/54616828)*120*(480+13))-640=283
+  * Example with TARGET_FPS of 30 fps for 1600x1400
+  * HBLANK=1/((1/204813108)*30*(1400+13))-1600=3232
+  */
+
+#define MIRA220_HBLANK_640x480_120FPS 3860
+#define MIRA220_HBLANK_1600x1400_30FPS 2900
+#define MIRA220_HBLANK_1600x1400_1500 1400
+
+#define MIRA220_HBLANK_400x400_304 2640
+#define MIRA220_HBLANK_640x480_304 2400
+#define MIRA220_HBLANK_1600x1400_304 1440
+
+#define MIRA220_REG_TEST_PATTERN CCI_REG8(0x2091)
+#define MIRA220_TEST_PATTERN_DISABLE 0x00
+#define MIRA220_TEST_PATTERN_VERTICAL_GRADIENT 0x01
+
+/* Embedded metadata stream structure */
+#define MIRA220_EMBEDDED_LINE_WIDTH 16384
+#define MIRA220_NUM_EMBEDDED_LINES 1
```

Comment:
You'd said in https://github.com/raspberrypi/linux/pull/6717#discussion_r2025643185 "I have removed the embedded data", so why are these still here. Likewise METADATA_PAD at line 158.